src/Microsoft.ML.FastTree/GamTrainer.cs

```diff
@@ -1048,6 +1049,40 @@ public void SaveSummary(TextWriter writer, RoleMappedSchema schema)
             ((ICanSaveInTextFormat)this).SaveAsText(writer, schema);
         }
 
+        public ValueMapper<TSrc, VBuffer<float>> GetFeatureContributionMapper<TSrc, TDstContributions>
+            (int top, int bottom, bool normalize)
+        {
+            Contracts.Check(typeof(TSrc) == typeof(VBuffer<float>));
+            Contracts.Check(typeof(TDstContributions) == typeof(VBuffer<float>));
+
+            ValueMapper<VBuffer<float>, VBuffer<float>> del =
+                (in VBuffer<float> srcFeatures, ref VBuffer<float> dstContributions) =>
+                {
+                    GetFeatureContributions(in srcFeatures, ref dstContributions, top, bottom, normalize);
+                };
+            return (ValueMapper<TSrc, VBuffer<float>>)(Delegate)del;
+        }
+
+        private void GetFeatureContributions(in VBuffer<float> features, ref VBuffer<float> contributions,
+                        int top, int bottom, bool normalize)
```

Comment:
Are you using the function from line 866: internal void GetFeatureContributions(in VBuffer<float> features, ref VBuffer<float> contribs, ref BufferBuilder<float> builder) 

If not, why not?

Reply:
Artidoro, I didn't see that one, but we should only keep one version of GetFeatureContributions. 

@rogancarr, In my version (line 1066) I'm using VBufferEditor and always use dense value of features, b/c the feature contributions could be significant even for features with value 0. So I didn't bother with dense/sparse checks, and the code is much shorter. I also have the top/bottom logic. I think you added the version on line 866, which uses the Intercept value and checks for dense/sparse.

Let's discuss which version to keep.

---
In reply to: [240429834](https://github.com/dotnet/machinelearning/pull/1856#discussion_r240429834) [](ancestors = 240429834)

Reply:
On the topic of multiplicity of functions, the `GetFeatureContributions` with this signature:
```cs
internal void GetFeatureContributions(in VBuffer<float> features, ref VBuffer<float> contribs, ref BufferBuilder<float> builder)
```

Is used by an internal tool. Let's leave it in place for now until we can remove it. I'm okay adding a separate method with the same name in that we use for `FeatureContributionCalculator` calls.

Reply:
Ignore the previous comment. You can get rid of the `BufferBuilder` implementation.

Reply:
On sparsity: I agree with this sentiment. It can be important that a feature is missing (zero).

Reply:
However, now that I've looked over the implementations, my suggestion would be to use the existing implementation, but
1. Change to using the `VBufferEditor`
2. Drop the `Intercept` term: This doesn't make sense
3. Change the method signature to match the new method signature.
4. Add the `SparsifyNormalize` call at the end.

The reason is that the sparsity handling of the existing function is nice because it doesn't densify the array to iterate over it and instead keeps track of the indices.

Reply:
As @shmoradims pointed out, `DenseValues()` is an iterator, so this should work as-is. Let's delete the old implementation then.

Reply:
Deleted the old code.

---
In reply to: [241154932](https://github.com/dotnet/machinelearning/pull/1856#discussion_r241154932) [](ancestors = 241154932)